### PR TITLE
Add ability to use GEFS replay ICs

### PR DESCRIPTION
**Pull request by NeilBarton-NOAA**
# Description
The PR allows the use of ICs from PSL's replay analysis. These replay ICs will be used for GEFS reforecasting and SFS. Two main changes are associated with these updates: (1) replay ICs being valid at 3Z, and (2) the use of warm starts. 

This addresses https://github.com/NOAA-EMC/global-workflow/issues/1838 

# Type of change

- New feature 

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- GEFS Forecast-only on Hera using replay ICs
- GEFS Forecast-only on Hera using warm start ICs
- GEFS Forecast-only on Hera using cold start ICs
- GFS Forecast-only on Hera CI test

# Checklist
- [X] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: scripts/exgfs_wave_post_pnt.sh
Comment (WalterKolczynski-NOAA) [inline scripts/exgfs_wave_post_pnt.sh]: Please retain the use of `${NLN}`

File: scripts/exglobal_stage_ic.sh
Comment (WalterKolczynski-NOAA) [inline scripts/exglobal_stage_ic.sh]:
```suggestion
    prev_atmos_copy_list=(fv_core.res.nc)
    if [[ "${USE_REPLAY_ICS:-"false"}" == "false" ]]; then
      prev_atmos_copy_list+=(coupler.res)
    fi
    for ftype in "${prev_atmos_copy_list[@]}"; do
      src="${BASE_CPLIC}/${CPL_ATMIC:-}/${PDY}${cyc}/${MEMDIR}/atmos/${PDY}.${cyc}0000.${ftype}"
      tgt="${COM_ATMOS_RESTART_PREV}/${PDY}.${cyc}0000.${ftype}"
      ${NCP} "${src}" "${tgt}"
      rc=$?
      ((rc != 0)) && error_message "${src}" "${tgt}" "${rc}"
      err=$((err + rc))
    done
```

File: scripts/exglobal_stage_ic.sh
Comment (WalterKolczynski-NOAA) [inline scripts/exglobal_stage_ic.sh]: If this is now going to use the previous cycle's date, the variable name should be changed to `COM_WAVE_RESTART_PREV` to be consistent.

Also, I believe this will break the existing GEFS test.

File: scripts/exglobal_stage_ic.sh
Comment (WalterKolczynski-NOAA) [inline scripts/exglobal_stage_ic.sh]: Why are atmosphere perturbations coming from the ocean IC directory? Doesn't matter right now since they are identical, but this should use `CPL_ATMIC`.

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]: Please indent this block (except the EOF)

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]: Please indent this block.

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]: Please indent this block.

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]: Please indent this block.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
Does this resolve #1838 or is it only part of it? The details were never filled in to that issue to know what needs to be done.

File: scripts/exglobal_stage_ic.sh
Comment (NeilBarton-NOAA) [reply scripts/exglobal_stage_ic.sh]: Error in coping and pasting. Updated

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests. There may also be conflicts with #2545 that will need to be resolved.

**Comment by WalterKolczynski-NOAA** [comment]
Didn't notice pynorms was failing. Please remove the trailing whitespace in `gefs_tasks.py`:
```
./workflow/rocoto/gefs_tasks.py:12:43: W291 trailing whitespace
./workflow/rocoto/gefs_tasks.py:13:18: W291 trailing whitespace
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Didn't notice pynorms was failing. Please remove the trailing whitespace in `gefs_tasks.py`:
```
./workflow/rocoto/gefs_tasks.py:12:43: W291 trailing whitespace
./workflow/rocoto/gefs_tasks.py:13:18: W291 trailing whitespace
```

**Comment by NeilBarton-NOAA** [comment]
updated 

> Didn't notice pynorms was failing. Please remove the trailing whitespace in `gefs_tasks.py`:
> 
> ```
> ./workflow/rocoto/gefs_tasks.py:12:43: W291 trailing whitespace
> ./workflow/rocoto/gefs_tasks.py:13:18: W291 trailing whitespace
> ```

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA @aerorahul I merged develop into this branch and changed to use files named with the 03 start. In my opinion, it's ready to comments and testing. I tested the GEFS CI and currently testing a GFS forecast.

File: parm/config/gefs/config.base
Comment (JessicaMeixner-NOAA) [inline parm/config/gefs/config.base]: I think this needs to be defined for the GFS configs.

Comment (NeilBarton-NOAA) [reply parm/config/gefs/config.base]: thanks, added

Comment (JessicaMeixner-NOAA) [reply parm/config/gefs/config.base]: @NeilBarton-NOAA I don't see this in parm/config/gfs/config.base, maybe it didn't get pushed?

**Comment by NeilBarton-NOAA** [comment]
@JessicaMeixner-NOAA I added the update to the wrong branch. It is update/merged now. Thanks.

**Comment by JessicaMeixner-NOAA** [comment]
> @JessicaMeixner-NOAA I added the update to the wrong branch. It is update/merged now. Thanks.

Thanks @NeilBarton-NOAA!

File: parm/config/gefs/config.wave
Comment (WalterKolczynski-NOAA) [inline parm/config/gefs/config.wave]: Note: #2491 is moving this to `config.base`, so we'll need to make sure this change is included no matter what the order of these two PRs.

File: ush/wave_outp_spec.sh
Comment (WalterKolczynski-NOAA) [inline ush/wave_outp_spec.sh]: We're trying to eliminate `CDATE`, so might as well make this correct from the start.
```suggestion
    model_start_date="${PDY}${cyc}"
```

File: workflow/rocoto/gefs_tasks.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/gefs_tasks.py]: This would fail if used on an experiment of more than one cycle. Note how the non-replay case is using the rocoto `<cyclestr>` formatting (`@Y`, `@m`, etc.). The replay needs to do the same instead of using `SDATE` (see also next comment).
```suggestion
        dtg_prefix = "@Y@m@d.@H0000"
        if self._configs['base']['END_OF_IAU_START']:
            offset = "+03:00:00"
        else:
            offset = None
```

File: workflow/rocoto/gefs_tasks.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/gefs_tasks.py]: The `add_dependency()` method can already handle these complicated offset situations (we need to add an offset to the filename but not the directory). We just need to pass an array of data and offsets and they will be combined:
```suggestion
                    data = [prefix, f"{dtg_prefix}.{file}"]
                    dep_dict = {'type': 'data', 'data': data, 'offset': [None, offset]}
                    deps.append(rocoto.add_dependency(dep_dict))
                prefix = f"{cpl_ic['BASE_CPLIC']}/{cpl_ic['CPL_ATMIC']}/@Y@m@d@H/mem000/med"
                data = [prefix, f"{dtg_prefix}.ufs.cpld.cpl.r.nc"]
                dep_dict = {'type': 'data', 'data': data, 'offset': [None, offset]}
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: ush/wave_outp_spec.sh
Comment (NeilBarton-NOAA) [reply ush/wave_outp_spec.sh]: edited

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA 

There are some errors in the xml file with the updates to workflow/rocoto/gefs_tasks.py
 
Error: Element workflow failed to validate content at

Could you help debug?

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA
> 
> There are some errors in the xml file with the updates to workflow/rocoto/gefs_tasks.py
> 
> Error: Element workflow failed to validate content at
> 
> Could you help debug?

Sorry, was busy much of the day. If you give me a path, I can take a look, but it often means you have a task dependency for a task name that doesn't exist or some other task missing/misspelled issue.

Edit: although since we were mostly mucking with offsets in these changes, it's probably that. I'll need to see what the rendered XML looks like.

**Comment by WalterKolczynski-NOAA** [comment]
One error I see for sure is you didn't update the other instances of `dtg_prefix` to be lowercase in `gefs_tasks.py`.

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA The above error is associated with something wrong in the xml file, not dtg_prefix. I just uploaded the lower case dtg_prefix so the error can be reproduced.

File: parm/config/gfs/config.base
Comment (aerorahul) [inline parm/config/gfs/config.base]: A well-rounded way would be to define an `offset` from `current_cycle`.  For replay ICs, the offset is `+3h`.  For cycling/forecast-only with IAU the offset is `-3h`.  If offset is `0h`, the warm starts can be expected to be valid at the `current_cycle`.

File: scripts/exglobal_stage_ic.sh
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh]: formulate in terms of `offset`.

File: workflow/rocoto/gefs_tasks.py
Comment (aerorahul) [inline workflow/rocoto/gefs_tasks.py]: This should be handled better. Hard-wiring offsets deep in this code will make this difficult to maintain.
This is a capability that can also be used in GFS if someone wants to start a forecast from the end of the assimilation window e.g. a reanalysis

File: parm/config/gefs/config.fcst
Comment (WalterKolczynski-NOAA) [reply parm/config/gefs/config.fcst]:
```suggestion
if (( OFFSET_START_HOUR != 0 )); then
```

File: scripts/exglobal_stage_ic.sh
Comment (WalterKolczynski-NOAA) [reply scripts/exglobal_stage_ic.sh]:
```suggestion
    if (( OFFSET_START_HOUR == 0 )); then
```

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]:
```suggestion
    if (( OFFSET_START_HOUR == 0 )); then
```

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [reply ush/forecast_postdet.sh]:
```suggestion
      if (( OFFSET_START_HOUR != 0 )); then
```

File: ush/forecast_predet.sh
Comment (WalterKolczynski-NOAA) [reply ush/forecast_predet.sh]:
```suggestion
  if (( OFFSET_START_HOUR != 0 )); then
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Sorry for the delay, been a busy kind of week. Also waiting on a CI fix to get in. Just the linter suggestions left, I think.

**Comment by WalterKolczynski-NOAA** [comment]
@NeilBarton-NOAA Please merge in develop and fix any conflicts, and then I think this is ready for CI testing.

File: ush/forecast_det.sh
Comment (aerorahul) [inline ush/forecast_det.sh]: This file is not present when `APP=ATM`, so its not an indicator for warm start.
Besides, one can "warm" start the atm, ocean and ice without this file if set correctly in the ufs_configure (I think).

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

File: ush/forecast_det.sh
Comment (JessicaMeixner-NOAA) [reply ush/forecast_det.sh]: if you set it the mediator to a cold start in the ufs_confugire, you should be able to warm start everything but the mediator.

**Comment by JessicaMeixner-NOAA** [comment]
I don't see any issues with the updates made for the wave scripts, which is assuming why my review was asked for.  Also tagging @sbanihash for her awareness of wave script changes.    I didn't make a GFS run to triple check, but visual inspection does not raise any red-flags for any of the updates here for waves.

**Comment by WalterKolczynski-NOAA** [comment]
@NeilBarton-NOAA If you resolve the conflicts, I think this is ready to test.

File: ush/forecast_det.sh
Comment (NeilBarton-NOAA) [reply ush/forecast_det.sh]: updated

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA updated and merged with dev

**Comment by WalterKolczynski-NOAA** [comment]
Looks like multiple issues. One is that the dependency path for wave restart files in stage_ic is missing a slash:

```
  dependencies
    AND is not satisfied
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_ctrl.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile1.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile2.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile3.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile4.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile5.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile6.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile1.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile2.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile3.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile4.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile5.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile6.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/ocean/20210323.120000.MOM.res.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/ice/20210323.120000.cice_model.res.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/wave20210323.120000.restart.glo_500 does not exist
```

**Comment by WalterKolczynski-NOAA** [comment]
Issue the second is stmp filled on `/work2` again:

```
hercules-login-1 ~/>df /work2/noaa/stmp
Filesystem                                                                            1K-blocks         Used Available Use% Mounted on
172.27.100.53@o2ib:172.27.100.54@o2ib:172.27.100.55@o2ib:172.27.100.56@o2ib:/data2 358612992000 358612992000         0 100% /work2
```

**Comment by WalterKolczynski-NOAA** [comment]
@NeilBarton-NOAA is this ready for another try?

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA yes, I tested the 3Dvar ATM CI last night

**Comment by WalterKolczynski-NOAA** [comment]
@NeilBarton-NOAA Setup experiment failed for the GEFS test. Something is wrong in the `stage_ic` task:
```
2024-06-11 18:29:16,135 - DEBUG    - root        : setup_xml.py /scratch1/NCEPDEV/global/CI/2559/RUNTESTS/EXPDIR/C48_S2SWA_gefs_f0410de9
Finalizing initialize
sourcing config.stage_ic
sourcing config.fcst
sourcing config.atmos_products
sourcing config.efcs
sourcing config.atmos_ensstat
sourcing config.waveinit
sourcing config.wavepostsbs
sourcing config.wavepostpnt
sourcing config.oceanice_products
sourcing config.prep_emissions
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/tasks.py", line 239, in get_task
    return getattr(self, task_name, *args, **kwargs)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/gefs_tasks.py", line 16, in stage_ic
    offset = str(self._configs['base']['OFFSET_START_HOUR']).zfill(2) + ":00:00"
                 ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/wxflow/attrdict.py", line 82, in __missing__
    if object.__getattribute__(self, '__frozen'):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'AttrDict' object has no attribute '__frozen'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2559/gefs//workflow/create_experiment.py", line 101, in <module>
    setup_xml.main(setup_xml_args)
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/setup_xml.py", line 73, in main
    xml = rocoto_xml_factory.create(f'{net}_{mode}', app_config, rocoto_param_dict)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/wxflow/factory.py", line 71, in create
    return self._builders[key](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/gefs_xml.py", line 14, in __init__
    super().__init__(app_config, rocoto_config)
  File "/scratch1/
```NCEPDEV/global/CI/2559/gefs/workflow/rocoto/workflow_xml.py", line 27, in __init__
    task_list = get_wf_tasks(app_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/workflow_tasks.py", line 21, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/tasks.py", line 241, in get_task
    raise AttributeError(f'"{task_name}" is not a valid task.\n' +
AttributeError: "stage_ic" is not a valid task.
Valid tasks are:
aerosol_init, stage_ic, prep, anal, sfcanl, analcalc, analdiag, arch, cleanup, prepatmiodaobs, atmanlinit, atmanlvar, atmanlfv3inc, atmanlfinal, prepoceanobs, ocnanalprep, ocnanalbmat, ocnanalrun, ocnanalecen, ocnanalchkpt, ocnanalpost, ocnanalvrfy, earc, ecen, echgres, ediag, efcs, eobs, eomg, epos, esfc, eupd, atmensanlinit, atmensanlletkf, atmensanlfv3inc, atmensanlfinal, aeroanlinit, aeroanlrun, aeroanlfinal, prepsnowobs, snowanl, fcst, atmanlupp, atmanlprod, atmupp, goesupp, atmosprod, oceanprod, iceprod, verfozn, verfrad, vminmon, metp, tracker, genesis, genesis_fsu, postsnd, awips_20km_1p0deg, fbwind, gempak, gempakmeta, gempakmetancdc, gempakncdcupapgif, gempakpgrb2spec, npoess_pgrb2_0p5degwaveawipsbulls, waveawipsgridded, wavegempak, waveinit, wavepostbndpnt, wavepostbndpntbll, wavepostpnt, wavepostsbs, waveprep, npoess, mos_stn_prep, mos_grd_prep, mos_ext_stn_prep, mos_ext_grd_prep, mos_stn_fcst, mos_grd_fcst, mos_ext_stn_fcst, mos_ext_grd_fcst, mos_stn_prdgen, mos_grd_prdgen, mos_ext_stn_prdgen, mos_ext_grd_prdgen, mos_wx_prdgen, mos_wx_ext_prdgen
```
(Ignore the "not a valid task" part; that is the error message whenever there is an exception in one of those.)

**Comment by NeilBarton-NOAA** [comment]
updated and the GEFS test (and SFS replay test) forecasts are both running on my side

**Comment by WalterKolczynski-NOAA** [comment]
> updated and the GEFS test (and SFS replay test) forecasts are both running on my side

K. I'm not starting any more CI on Hera right now until we get some space cleaned up under global, so there might be a slight delay.

**Pull request merged**
